rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs

```diff
@@ -297,37 +307,58 @@ async fn kyt_check_utxo(utxo: &Utxo) -> Result<UtxoCheckStatus, UpdateBalanceErr
-                DisplayOutpoint(&utxo.outpoint),
-                addresses.join(",")
-            );
-            Ok(UtxoCheckStatus::Tainted)
-        }
-        CheckTransactionResponse::Passed => Ok(UtxoCheckStatus::Clean),
-        CheckTransactionResponse::Unknown(status) => {
-            log!(
-                P1,
-                "The KYT provider is temporarily unavailable: {:?}",
-                status
-            );
-            Err(UpdateBalanceError::TemporarilyUnavailable(format!(
-                "The KYT provider is temporarily unavailable: {:?}",
-                status
-            )))
+    for i in 0..MAX_CHECK_TRANSACTION_RETRY {
+        match check_transaction(new_kyt_principal, utxo, CHECK_TRANSACTION_CYCLES_REQUIRED)
+            .await
+            .map_err(|call_err| {
+                UpdateBalanceError::TemporarilyUnavailable(format!(
+                    "Failed to call KYT canister: {}",
+                    call_err
+                ))
+            })? {
+            CheckTransactionResponse::Failed(addresses) => {
+                log!(
+                    P0,
+                    "Discovered a tainted UTXO {} (due to input addresses {}) for update_balance({:?}) call",
+                    DisplayOutpoint(&utxo.outpoint),
+                    addresses.join(","),
+                    args,
+                );
+                return Ok(UtxoCheckStatus::Tainted);
+            }
+            CheckTransactionResponse::Passed => return Ok(UtxoCheckStatus::Clean),
+            CheckTransactionResponse::Unknown(CheckTransactionStatus::NotEnoughCycles) => {
+                log!(
+                    P1,
+                    "The KYT canister requires more cycles, Remaining tries: {}",
+                    MAX_CHECK_TRANSACTION_RETRY - i - 1
+                );
+                continue;
+            }
+            CheckTransactionResponse::Unknown(CheckTransactionStatus::Retriable(status)) => {
+                log!(
+                    P1,
+                    "The KYT canister is temporarily unavailable: {:?}",
+                    status
+                );
+                return Err(UpdateBalanceError::TemporarilyUnavailable(format!(
+                    "The KYT canister is temporarily unavailable: {:?}",
+                    status
+                )));
+            }
+            CheckTransactionResponse::Unknown(CheckTransactionStatus::Error(error)) => {
+                log!(P1, "KYT error: {:?}", error);
+                return Err(UpdateBalanceError::GenericError {
+                    error_code: ErrorCode::KytError as u64,
+                    error_message: format!("KYT error: {:?}", error),
+                });
+            }
         }
     }
+    Err(UpdateBalanceError::GenericError {
+        error_code: ErrorCode::KytError as u64,
+        error_message: format!("The KYT canister required too many calls to check_transaction"),
+    })
```

Comment:
Thanks for the fix! 2 follow-up questions:

1. In that case a user would need to call `update_balance` again to retry, correct?
2. Would it be possible to have a pocket IC test for this scenario (minter needs to try at least twice to validate a transaction)?

Reply:
> 1. In that case a user would need to call update_balance again to retry, correct?

Either that or the user can contact DFINITY to report this error. It would certainly be an anomaly. 

> 2. Would it be possible to have a pocket IC test for this scenario (minter needs to try at least twice to validate a transaction)?

The KYT canister has unit tests that where the caller would need to call more than once, but not pocket IC based tests. We will need to tweak some setup in order to trigger this scenario, and in particular, a much lower `CHECK_TRANSACTION_CYCLES_REQUIRED` is required. Maybe not in this PR though. I created a Jira ticket to track this https://dfinity.atlassian.net/browse/XC-241